clients/rust/tests/initialize_stake.rs

```diff
@@ -80,3 +71,287 @@ async fn initialize_config_with_mint_and_token() {
     assert_eq!(stake_account.validator, validator);
     assert_eq!(stake_account.authority, validator);
 }
+
+#[tokio::test]
+async fn fail_initialize_stake_with_initialized_account() {
+    let mut context = ProgramTest::new("stake_program", paladin_stake::ID, None)
+        .start_with_context()
+        .await;
+
+    // Given a config account and a validator's vote account.
+
+    let config = create_config(&mut context).await;
+    let validator = Pubkey::new_unique();
+    let vote = create_vote_account(&mut context, &validator, &validator);
+
+    // And we initialize the stake account.
+
+    let (stake_pda, _) = find_stake_pda(&validator, &config);
+
+    let transfer_ix = system_instruction::transfer(
+        &context.payer.pubkey(),
+        &stake_pda,
+        context
+            .banks_client
+            .get_rent()
+            .await
+            .unwrap()
+            .minimum_balance(Stake::LEN),
+    );
+
```

Comment:
This will guard from pre-allocated (or initialized) accounts, right? The program asserts that `stake.data_is_empty()` before trying to initialize the account – this would be covered by the `fail_initialize_stake_with_initialized_account` test.

I think this "manual" initialization can be added to other instructions where we need to assert that the account is initialized, so the discriminator must match.

Reply:
> The program asserts that `stake.data_is_empty()` before trying to initialize the account – this would be covered by the `fail_initialize_stake_with_initialized_account` test.

`fail_initialize_stake_with_initialized_account` is the test I left this comment on. Sorry, I think I maybe grabbed the wrong lines here - it should have been lower down.

I'm basically suggesting that instead of using the `InitializeStake` instruction to initialize the account, then using `InitializeStake` again to verify it won't re-initialize, you could set up the account with arbitrary data, to ensure that data initialized without the `Stake` discriminator (or even all zeroes) still fails on `stake.data_is_empty()`.

As I type it out, maybe it's not all that useful, but I'll leave it up to you.

Reply:
I will add it to another instruction, one that does not have the `stake.data_is_empty()` so we assert against the discriminator.